File: branches/Medusa/Source/Fantasy.Jobs/Foreach.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Fantasy.XSerialization;
using System.Text.RegularExpressions;

namespace Fantasy.Jobs
{
    [Instruction]
    [XSerializable("foreach", NamespaceUri = Consts.XNamespaceURI)]
    internal class Foreach : Sequence 
    {
        public override void Execute()
        {
            IJob job = this.Site.GetRequiredService<IJob>();
            IStringParser parser = this.Site.GetRequiredService<IStringParser>();
            Dictionary<string, object> context = new Dictionary<string, object>() 
            { 
              { "EnableTaskItemReader", false }
            };
            string s = parser.Parse(this.In, context);
            string[] strings = s != null ? s.Split(new char[]{';'}, StringSplitOptions.RemoveEmptyEntries) : new string[0];
            List<object> values = new List<object>();
            Regex reg = new Regex(@"^@\((?<cate>[\w]+)\)$");

            foreach (string str in strings)
            {
                Match match = reg.Match(str);
                if (match.Success)
                {
                    string cat = match.Groups["cate"].Value;
                    values.AddRange(job.GetEvaluatedItemsByCatetory(cat)); 
                }
                else
                {
                    values.Add(str);
                }
            }

            int index = job.RuntimeStatus.Local.GetValue("foreach.index", 0);
            while (index < values.Count)
            {
                object value = values[index];
                if ( ! (value is string) || (!string.IsNullOrWhiteSpace((string)value) || !_skipEmptyString))
                {
                    job.RuntimeStatus.Local[Var] = value;

                    this.ExecuteSequence(); 
                    this.ResetSequenceIndex();
                }
                index++;
                job.RuntimeStatus.Local["foreach.index"] = index;
            }
           

        }


        [XAttribute("var")] 
        public string Var { get; set; }

        [XAttribute("in")]
        public string In { get; set; }

        private bool _skipEmptyString = true;

        [XAttribute("skipEmptyString")]
        public bool SkipEmptyString
        {
            get { return _skipEmptyString; }
            set { _skipEmptyString = value; }
        }

        
    }
}

File: branches/Stheno/AutoConverted/Fantasy.Jobs/Foreach.java
package Fantasy.Jobs;

import Fantasy.XSerialization.*;

//C# TO JAVA CONVERTER TODO TASK: Java annotations will not correspond to .NET attributes:
//[Instruction, XSerializable("foreach", NamespaceUri = Consts.XNamespaceURI)]
public class Foreach extends Sequence
{
	@Override
	public void Execute()
	{
		IJob job = this.getSite().<IJob>GetRequiredService();
		IStringParser parser = this.getSite().<IStringParser>GetRequiredService();
		java.util.HashMap<String, Object> context = new java.util.HashMap<String, Object>();
		context.put("EnableTaskItemReader", false);
		String s = parser.Parse(this.In, context);
		String[] strings = s != null ? s.split(new char[]{';'}, StringSplitOptions.RemoveEmptyEntries) : new String[0];
		java.util.ArrayList<Object> values = new java.util.ArrayList<Object>();
		Regex reg = new Regex("^@\\((?<cate>[\\w]+)\\)$");

		for (String str : strings)
		{
			Match match = reg.Match(str);
			if (match.Success)
			{
				String cat = match.Groups["cate"].getValue();
				values.addAll(job.GetEvaluatedItemsByCatetory(cat));
			}
			else
			{
				values.add(str);
			}
		}

		int index = job.getRuntimeStatus().getLocal().GetValue("foreach.index", 0);
		while (index < values.size())
		{
			Object value = values.get(index);
			if (! (value instanceof String) || (!String.IsNullOrWhiteSpace((String)value) || !_skipEmptyString))
			{
				job.getRuntimeStatus().getLocal().setItem(Var, value);

				this.ExecuteSequence();
				this.ResetSequenceIndex();
			}
			index++;
			job.getRuntimeStatus().getLocal().setItem("foreach.index", index);
		}


	}


//C# TO JAVA CONVERTER TODO TASK: Java annotations will not correspond to .NET attributes:
	//[XAttribute("var")]
	public String Var = null;

//C# TO JAVA CONVERTER TODO TASK: Java annotations will not correspond to .NET attributes:
	//[XAttribute("in")]
	public String In = null;

//C# TO JAVA CONVERTER TODO TASK: Java annotations will not correspond to .NET attributes:
	//[XAttribute("skipEmptyString")]
	private boolean _skipEmptyString = true;

}